File: overlord/configstate/configcore/experimental_test.go
// -*- Mode: Go; indent-tabs-mode: t -*-

/*
 * Copyright (C) 2018 Canonical Ltd
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 3 as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

package configcore_test

import (
	"fmt"
	"path/filepath"

	. "gopkg.in/check.v1"

	"github.com/snapcore/snapd/features"
	"github.com/snapcore/snapd/osutil"
	"github.com/snapcore/snapd/overlord/configstate/configcore"
	"github.com/snapcore/snapd/testutil"
)

type experimentalSuite struct {
	configcoreSuite
}

var _ = Suite(&experimentalSuite{})

func featureConf(feature features.SnapdFeature) string {
	return "experimental." + feature.String()
}

func (s *experimentalSuite) TestConfigureExperimentalSettingsInvalid(c *C) {
	for _, feature := range features.KnownFeatures() {
		conf := &mockConf{
			state:   s.state,
			changes: map[string]any{featureConf(feature): "foo"},
		}
		err := configcore.FilesystemOnlyRun(classicDev, conf)
		c.Check(err, ErrorMatches, fmt.Sprintf(`%s can only be set to 'true' or 'false'`, featureConf(feature)))
	}
}

func (s *experimentalSuite) TestConfigureExperimentalSettingsHappy(c *C) {
	for _, feature := range features.KnownFeatures() {
		for _, t := range []string{"true", "false"} {
			conf := &mockConf{
				state: s.state,
				conf:  map[string]any{featureConf(feature): t},
			}
			err := configcore.FilesystemOnlyRun(classicDev, conf)
			c.Check(err, IsNil)
		}
	}
}

func (s *experimentalSuite) TestExportedFeatures(c *C) {
	conf := &mockConf{
		state: s.state,
		conf:  map[string]any{featureConf(features.HiddenSnapDataHomeDir): true},
	}
	err := configcore.FilesystemOnlyRun(classicDev, conf)
	c.Assert(err, IsNil)
	c.Check(features.HiddenSnapDataHomeDir.ControlFile(), testutil.FilePresent)

	delete(conf.changes, "experimental.hidden-snap-folder")
	err = configcore.FilesystemOnlyRun(classicDev, conf)
	c.Assert(err, IsNil)
	c.Check(features.HiddenSnapDataHomeDir.ControlFile(), testutil.FilePresent)
}

func (s *experimentalSuite) TestFilesystemOnlyApply(c *C) {
	conf := configcore.PlainCoreConfig(map[string]any{
		"experimental.hidden-snap-folder": "true",
	})
	tmpDir := c.MkDir()
	c.Assert(configcore.FilesystemOnlyApply(classicDev, tmpDir, conf), IsNil)
	c.Check(osutil.FileExists(filepath.Join(tmpDir, "/var/lib/snapd/features/hidden-snap-folder")), Equals, true)
}

func (s *experimentalSuite) TestFilesystemOnlyApplyValidationFails(c *C) {
	conf := configcore.PlainCoreConfig(map[string]any{
		"experimental.hidden-snap-folder": 1,
	})
	tmpDir := c.MkDir()
	c.Assert(configcore.FilesystemOnlyApply(classicDev, tmpDir, conf), ErrorMatches, `experimental.hidden-snap-folder can only be set to 'true' or 'false'`)
}

File: cmd/libsnap-confine-private/experimental-test.c
/*
 * Copyright (C) 2018 Canonical Ltd
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 3 as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

#include "experimental.h"
#include "experimental.c"

#include <limits.h>

#include <glib.h>

#include "string-utils.h"
#include "test-utils.h"

static char *sc_testdir(void)
{
	char *d = g_dir_make_tmp(NULL, NULL);
	g_assert_nonnull(d);
	g_test_queue_free(d);
	g_test_queue_destroy((GDestroyNotify) rm_rf_tmp, d);
	return d;
}

// Set the feature flag directory to given value, useful for cleanup handlers.
static void set_feature_flag_dir(const char *dir)
{
	feature_flag_dir = dir;
}

// Mock the location of the feature flag directory.
static void sc_mock_feature_flag_dir(const char *d)
{
	g_test_queue_destroy((GDestroyNotify) set_feature_flag_dir,
			     (void *)feature_flag_dir);
	set_feature_flag_dir(d);
}

static void test_experimental_flag_active__missing_dir(void)
{
	const char *d = sc_testdir();
	char subd[PATH_MAX];
	sc_must_snprintf(subd, sizeof subd, "%s/absent", d);
	sc_mock_feature_flag_dir(subd);
	g_assert(!sc_experimental_flag_active(SC_PER_USER_MOUNT_NAMESPACE));
}

static void test_experimental_flag_active__missing_file(void)
{
	const char *d = sc_testdir();
	sc_mock_feature_flag_dir(d);
	g_assert(!sc_experimental_flag_active(SC_PER_USER_MOUNT_NAMESPACE));
}

static void test_experimental_flag_active__present_file(void)
{
	const char *d = sc_testdir();
	sc_mock_feature_flag_dir(d);
	char pname[PATH_MAX];
	sc_must_snprintf(pname, sizeof pname, "%s/per-user-mount-namespace", d);
	g_file_set_contents(pname, "", -1, NULL);

	g_assert(sc_experimental_flag_active(SC_PER_USER_MOUNT_NAMESPACE));
}

static void __attribute__ ((constructor)) init(void)
{
	g_test_add_func("/experimental/missing_dir",
			test_experimental_flag_active__missing_dir);
	g_test_add_func("/experimental/missing_file",
			test_experimental_flag_active__missing_file);
	g_test_add_func("/experimental/present_file",
			test_experimental_flag_active__present_file);
}
